remoulade/composition.py

```diff
@@ -14,48 +14,118 @@
+        return GroupInfo(group_id=group_id, count=count, results=results)
 
 
 class pipeline:
     """Chain actors together, passing the result of one actor to the
     next one in line.
 
     Parameters:
-      children(Iterator[Message|pipeline]): A sequence of messages or
-        pipelines.  Child pipelines are flattened into the resulting
+      children(Iterator[Message|pipeline|group]): A sequence of messages or
+        pipelines or groups.  Child pipelines are flattened into the resulting
         pipeline.
       broker(Broker): The broker to run the pipeline on.  Defaults to
         the current global broker.
+
+    Attributes:
+        children(List[Message|group]) The sequence of messages or groups to execute as a pipeline
     """
 
     def __init__(self, children, *, broker=None):
         self.broker = broker or get_broker()
-        self.messages = messages = []
 
+        self.children = []
         for child in children:
             if isinstance(child, pipeline):
-                for message in child.messages:
-                    messages.append(message.copy())
+                self.children += child.children
+            elif isinstance(child, group):
+                self.children.append(child)
             else:
-                messages.append(child.copy())
+                self.children.append(child.copy())
 
-        for message, next_message in zip(messages, messages[1:]):
-            message.options["pipe_target"] = next_message.asdict()
+    def _build(self, *, last_options=None):
+        """ Build the pipeline, return the first message to be enqueued or integrated in another pipeline
+
+        Build the pipeline by starting at the end. We build a message with all it's options in one step and
+        we serialize it (asdict) as the previous message pipe_target in the next step.
+
+        We need to know what is the options (pipe_target) of the pipeline before building it because we cannot
+        edit the pipeline after it has been built.
+
+        Parameters:
+            last_options(dict): options to be assigned to the last actor of the pipeline (ex: pipe_target)
+
+        Returns:
+            the first message of the pipeline
+        """
+        next_child = None
+        for child in reversed(self.children):
+            if next_child:
+                if isinstance(next_child, list):
+                    options = {'pipe_target': [m.asdict() for m in next_child]}
+                else:
+                    options = {'pipe_target': next_child.asdict()}
+            else:
+                options = last_options or {}
+
+            if isinstance(child, group):
+                options = {'group_info': child.info.asdict(), **options}
+                messages = []
+                for group_child in child.children:
+                    if isinstance(group_child, pipeline):
+                        first = group_child._build(last_options=options)
```

Comment:
This is a bit surprising at first sight, using a method on a child of a child...